src/jit/morph.cpp

```diff
@@ -10648,13 +10650,44 @@ GenTree* Compiler::fgMorphCopyBlock(GenTree* tree)
                     CORINFO_CLASS_HANDLE classHnd = lvaTable[destLclNum].lvVerTypeInfo.GetClassHandle();
                     CORINFO_FIELD_HANDLE fieldHnd =
                         info.compCompHnd->getFieldInClass(classHnd, lvaTable[fieldLclNum].lvFldOrdinal);
-                    curFieldSeq = GetFieldSeqStore()->CreateSingleton(fieldHnd);
+                    curFieldSeq        = GetFieldSeqStore()->CreateSingleton(fieldHnd);
+                    var_types destType = lvaGetDesc(fieldLclNum)->lvType;
 
-                    src = gtNewOperNode(GT_ADD, TYP_BYREF, src,
-                                        new (this, GT_CNS_INT)
-                                            GenTreeIntCon(TYP_I_IMPL, lvaTable[fieldLclNum].lvFldOffset, curFieldSeq));
-
-                    src = gtNewIndir(lvaTable[fieldLclNum].TypeGet(), src);
+                    bool done = false;
+                    if (lvaGetDesc(fieldLclNum)->lvFldOffset == 0)
```

Comment:
Just wondering, do we know why this wasn't showing on x64 (and was only showing on x86, Arm, and Arm64)?

Reply:
... and now we know. It's still failing on arm64.

Reply:
The arm64 failure looks like a related but slightly different issue. For some reason in the double case the SIMD16 lclVar that's partially written in `ToVector128` is not marked as address-taken. I'd like to investigate further before deciding whether to incorporate a fix for that issue into this PR.

Reply:
> For some reason in the double case the SIMD16 lclVar that's partially written in ToVector128 is not marked as address-taken.

Address taken or not enregistered? I'm on the road now so I can't check easily but I'm pretty sure LocalAddressVisitor won't flag such a variable as address exposed. AFAIR it is expected that morph will convert such a tree to a LCL_FLD rather than keeping it as address taken. But then of course, we're talking about SIMD so it wouldn't be too surprising that a SIMD specific case was missed somewhere.

Reply:
The fundamental issue is that codegen expects it to *not* be enregistered, which in this context I would expect to have happened by marking it as address-taken. As you say, it seems to be related to the SIMD-ness - i.e. a SIMD store/assignment is not being treated in the same way as a primitive type.

However, it's Thanksgiving so I'll plan to investigate this more on Monday!

Reply:
@CarolEidt It looks like this is one of the cases we were discussing in the "remove GT_ASG" issue. Because `LCL_FLD` doesn't carry enough info about structs it cannot be used as the destination of a struct copy so instead we get a `BLK(ADDR(LCL_VAR))` tree. For such a tree, it is normal to not have the lclvar address-exposed, it's even normal for the `LCL_VAR` node to act as a SSA def.

This case also occurs when you're assigning to a part of a large, unpromotable struct, even on x64, and it works fine.

The actual problem might be somewhere in ARM64 lowering, the `LCL_VAR_ADDR` node that appears after rationalization might need to be contained so codegen doesn't actually try to obtain the address of the lclvar. But it's not:
```
N001 (  3,  2) [000017] -c----------        t17 =    LCL_VAR   struct(P) V11 loc1         
                                                  *    long   V11._00 (offs=0x00) -> V13 tmp1          $140
N002 (  1,  1) [000026] U------N----        t26 =    LCL_VAR_ADDR byref  V10 loc0         ud:2->3
                                                 /--*  t26    byref  
                                                 +--*  t17    struct 
N004 (  4,  4) [000018] nA----------              *  STORE_BLK(8) struct (copy) (Unroll)
```